controllers/nutanixmachine_controller.go

```diff
@@ -514,6 +521,33 @@ func (r *NutanixMachineReconciler) reconcileNormal(rctx *nctx.MachineContext) (r
 	return reconcile.Result{}, nil
 }
 
+// checkFailureDomainStatus checks and sets the NutanixMachine.status.failureDomain if necessary
+func (r *NutanixMachineReconciler) checkFailureDomainStatus(rctx *nctx.MachineContext) error {
+	if rctx.Machine.Spec.FailureDomain == nil || *rctx.Machine.Spec.FailureDomain == "" {
+		return nil
+	}
+
+	// Fetch the referent failure domain object
+	fdObj := &infrav1.NutanixFailureDomain{}
+	fdName := *rctx.Machine.Spec.FailureDomain
+	fdKey := client.ObjectKey{Name: fdName, Namespace: rctx.NutanixMachine.Namespace}
+	err := r.Get(rctx.Context, fdKey, fdObj)
```

Comment:
We already have this code inside the `validateMachineConfig` function, where the new FD changes are mutating the NutanixMachine.
https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/cd2407f11b24a785ce6df1f67646c001b939beb3/controllers/nutanixmachine_controller.go#L557

What do you think about combining the 2 in a new function like `reconcileFailureDomain` that is always called before? https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/cd2407f11b24a785ce6df1f67646c001b939beb3/controllers/nutanixmachine_controller.go#L484

This has the benefit of having the status be set in a single place and validation to protect against unintended changes to the Machine's spec.

Reply:
validateMachineConfig() and validateFailureDomainRef() are only called when it needs to create the VM for the machine. While checkFailureDomainStatus() need to call for every reconcileNormal call. We can move these 4 lines of code to a separate function fetchFailureDomainObj(). But since it's just 4 lines of code, it is not much difference.

Reply:
@yanhua121 I think `validateFailureDomainRef()` is pretty unexpected now that it is mutating the objects, thats why I suggested to refactor to a single place.

Reply:
We can refactor the FD related code in the controller to make it more readable, in a separate PR later.